File: data-structures/set/python/Set.py
"""
Author: Jennifer Chukwu
"""

class Entry:
    """
    This represents the template for the node object that is used for the linkedlist.
    """
        
    def __init__(self, element):
        """
        :param element: This represents the element contained in the Node.
        :type element: This element can be of any type ranging from strings, numbers, arrays etc.
        """
        self.element = element
        self.next = None

class Set:
    def __init__(self):

        # Prime used to increase chances of generating unique keys
        self.PRIME = 31
        self.entries = [None] * 29

    def add(self, element):
        """
        The `add` method Stores the element in the set.

        :param element: element to be stored in the set
        :return true if element to be added doesn't exist, and false if it does
        """

        # Generate hash (or index)
        hash = self.get_hash_for_element(element)

        # Store element, if there's no collision
        if self.entries[hash] is None:
            self.entries[hash] = Entry(element)
            return True
        
        # If there's a collision, insert entry at tail of bucket (linkedlist)
        else:
            current = self.entries[hash]
            while current.next is not None:
                if current.element == element:
                    return False
                current = current.next

            # If element exists already, return because we don't want duplicates
            if current.element == element:
                return False
            current.next = Entry(element)
            return True

    def contains(self, element):
        """
        The `contains` method checks if the element is in the set.

        :param element: element whose presence is to be checked in the set
        :return true if element exists, and false if it doesn't
        """

        # Generate hash (or index)
        hash = self.get_hash_for_element(element)
        current = self.entries[hash]

        # traverse bucket(linkedlist) to find out if the element exists
        while current is not None:
            if current.element == element:
                return True
            current = current.next

        # If element is not found, return false
        return False

    def list(self):
        """
        :return a list of all elements in the set
        """
        result = []
        for entry in self.entries:
            current = entry

            # Traverse every index in array
            while current is not None:
                result.append(current.element)
                current = current.next
        return result


    def remove(self, element):
        """
        The `remove` method removes element from the set.

        :param element: element to be removed
        :return true if element was removed, and false if it wasn't
        """

        # Generate hash
        hash = self.get_hash_for_element(element)
        current = self.entries[hash]

        # keep track of the previous element 
        prev = None

        # Traverse the list to find element 
        while current is not None:
            if current.element == element:
                if prev is None:
                    self.entries[hash] = current.next
                else:
                    prev.next = current.next
                return True
            prev = current
            current = current.next

        # If element doesn't exist, so was not deleted
        return False

    def get_hash_for_element(self, element):
        return (hash(element) * self.PRIME) % len(self.entries)

if __name__ == '__main__':
    my_set = Set()
    
    # Adding elements to the set
    print("Adding elements:")
    print("Add 10:", my_set.add(10))  # Should print True
    print("Add 20:", my_set.add(20))  # Should print True
    print("Add 30:", my_set.add(30))  # Should print True
    print("Add 10 again:", my_set.add(10))  # Should print False (already added)
    
    # Checking if elements are in the set
    print("\nChecking elements:")
    print("Contains 10:", my_set.contains(10))  # Should print True
    print("Contains 20:", my_set.contains(20))  # Should print True
    print("Contains 40:", my_set.contains(40))  # Should print False
    
    # Removing an element
    print("\nRemoving elements:")
    print("Remove 10:", my_set.remove(10))  # Should print True
    print("Remove 40:", my_set.remove(40))  # Should print False (not in set)
    
    # Listing all elements in the set
    print("\nCurrent set elements:")
    print(my_set.list())  # Should print [30, 20]

File: data-structures/set/cpp/Set.cpp
#include <iostream>
/**
 * @author: Jennifer Chukwu
 */

template <class T> class Set {
  /**
   * This represents the template for the object that is stored at the
   * underlying array's indexes
   */

private:
  struct Entry {
    T element;
    Entry *next;

    Entry(T element) : element(element), next(nullptr) {}
    Entry() : next(nullptr) {} // Default constructor to allow array of pointers
  };

  static const int TABLE_SIZE = 29; // Size of the hash table
  static const int PRIME = 31;      // Prime number used in the hash function
  Entry *entries[TABLE_SIZE];       // Array of pointers to Entry

  int getHashForElement(const T &element) {
    return (std::hash<T>()(element) * PRIME) % TABLE_SIZE;
  }

public:
  Set() {
    for (int i = 0; i < TABLE_SIZE; ++i) {
      entries[i] = nullptr; // Initialize all entries to nullptr
    }
  }

  ~Set() {
    for (int i = 0; i < TABLE_SIZE; ++i) {
      Entry *current = entries[i];
      while (current != nullptr) {
        Entry *toDelete = current;
        current = current->next;
        delete toDelete;
      }
    }
  }

  /**
   * Stores the element in the set.
   *
   * @param element element to be stored in the set
   * @return true if element to be added doesn't exist, and false if it does
   */
  bool add(T element) {
    // Generate hash (or index)
    int hash = getHashForElement(element);

    Entry *newEntry = new Entry(element);
    Entry *entry = entries[hash];

    // Store element, if there's no collision
    if (entry == nullptr) {
      entries[hash] = newEntry;
      return true;
    }

    // If there's a collision, insert entry at tail of bucket (linkedlist)
    while (entry != nullptr) {

      // If element exists already, return because we don't want duplicates
      if (element == entry->element) {
        return false;
      }

      entry = entry->next;
    }

    entry->next = newEntry;
    return true;
  }

  /**
   * Checks if the element is in the set.
   *
   * @param element element whose presence is to be checked in the set
   * @return true if element exists, and false if it doesn't
   */
  bool contains(T element) {
    // Generate hash (or index)
    int hash = getHashForElement(element);

    // If no element is stored at index, then return false
    if (entries[hash] == nullptr) {
      return false;
    }

    // If there's a collision, traverse bucket(linkedlist) to find
    // out if the element exists
    Entry *entry = entries[hash];
    while (entry != nullptr) {

      // If element is found, return true
      if (element == entry->element) {
        return true;
      }

      entry = entry->next;
    }

    // If element is not found, return false
    return false;
  }

  /**
   * @removes element from the set
   * @return true if element to be deleted was removed, and false if it wasn't
   */
  bool remove(const T &element) {
    // Generate hash
    int hash = getHashForElement(element);

    Entry *current = entries[hash];
    Entry *prev = nullptr;

    // traverse each valid entry
    while (current != nullptr) {
      if (current->element == element) {
        if (prev == nullptr) {
          entries[hash] = current->next;
        } else {
          prev->next = current->next;
        }
        delete current;
        // return true if element exists and was deleted
        return true;
      }
      prev = current;
      current = current->next;
    }
    // If element doesn't exist, so was not deleted
    return false;
  }

  /**
   * @outputs all the elements in the set
   */

  void traverse() {
    for (int i = 0; i < TABLE_SIZE; ++i) {
      Entry *current = entries[i];
      while (current != nullptr) {
        std::cout << current->element << ", ";
        current = current->next;
      }
    }
    std::cout << std::endl;
  }
};

int main() {
  Set<int> mySet;
  mySet.add(10);
  mySet.add(20);
  mySet.add(30);
  mySet.add(30);

  mySet.traverse();

  std::cout << "Contains 20: " << mySet.contains(20) << std::endl;
  std::cout << "Contains 20 after removal: " << mySet.remove(80) << std::endl;

  return 0;
}